util/generate_ast/_base.px

```diff
@@ -0,0 +1,45 @@
+class AST(mutable=ast.AST, metaclass=ImmutableMeta):
+    def __setattr__(self, attr, value):
+        if attr in self._fields and hasattr(self, attr):
+            raise AttributeError('Cannot modify ImmutableAST fields')
+        elif isinstance(value, (list, ast.AST)):
+            value = immutable(value)
+
+        self.__dict__[attr] = value
+
+    def __delattr__(self, attr):
+        if attr in self._fields:
+            raise AttributeError('Cannot modify ImmutableAST fields')
+        del self.__dict__[attr]
+
+    def __hash__(self):
+        try:
+            return self._hash_
+        except AttributeError:
+            pass
+
+        h = hash(type(self))
+        for _, n in iter_fields(self):
+            if isinstance(n, AST):
+                h += hash(n)
+            elif isinstance(n, tp.Sequence):
+                for c in n:
+                    h += hash(c)
+            else:
+                h += hash(n)
+        self._hash_ = h
+        return h
+
+    def __eq__(self, other):
+        if not isinstance(other, type(self)):
+            return NotImplemented
```

Comment:
Could this just be False?

Reply:
I guess this will try the other way (inverse) which may be okay too, this seems reasonable.